Update go version

pruthvitd opened the pull request:
**Upgrade ramen repository go version to v1.24**
	Updated following packages:
        1. go version to v1.24.X
        2. Controller-runtime - v0.21.0
        3. Kubernetes - v0.30.0
        4. golangci-lint - v2.2.2
        5. Updated mdl, ShellCheck, yamllint
	6. Resolved linter errors and deprecated warnings
        7. Updated other dependent packages
        
E2E tests are validated with updated version of go:
```
# ./run.sh -test.run TestValidation
2025-07-11T05:53:24.407-0700	INFO	Using config file "config.yaml"
2025-07-11T05:53:24.407-0700	INFO	Using log file "ramen-e2e.log"
2025-07-11T05:53:24.417-0700	INFO	Using "hub" cluster name: "hub"
2025-07-11T05:53:24.450-0700	INFO	Detected "c1" managed cluster name: "dr1"
2025-07-11T05:53:24.502-0700	INFO	Detected "c2" managed cluster name: "dr2"
2025-07-11T05:53:24.503-0700	INFO	Using DeployTimeout: 5m0s
2025-07-11T05:53:24.503-0700	INFO	Using UneployTimeout: 5m0s
2025-07-11T05:53:24.503-0700	INFO	Using EnableTimeout: 5m0s
2025-07-11T05:53:24.503-0700	INFO	Using DisableTimeout: 5m0s
2025-07-11T05:53:24.503-0700	INFO	Using FailoverTimeout: 15m0s
2025-07-11T05:53:24.503-0700	INFO	Using RelocateTimeout: 15m0s
2025-07-11T05:53:24.503-0700	INFO	Using RetryInterval: 5s
=== RUN   TestValidation
=== PAUSE TestValidation
=== CONT  TestValidation
2025-07-11T05:53:24.541-0700	INFO	Detected kubernetes distribution: "k8s"
2025-07-11T05:53:24.541-0700	INFO	Using namespaces: {RamenHubNamespace:ramen-system RamenDRClusterNamespace:ramen-system RamenOpsNamespace:ramen-ops ArgocdNamespace:argocd}
2025-07-11T05:53:24.622-0700	INFO	Validated clusters ["dr1", "dr2"] in DRPolicy "dr-policy"
2025-07-11T05:53:24.660-0700	INFO	Validated clusters ["dr1", "dr2"] in ClusterSet "default"
=== RUN   TestValidation/hub
=== PAUSE TestValidation/hub
=== RUN   TestValidation/c1
=== PAUSE TestValidation/c1
=== RUN   TestValidation/c2
=== PAUSE TestValidation/c2
=== CONT  TestValidation/hub
=== CONT  TestValidation/c2
=== CONT  TestValidation/c1
2025-07-11T05:53:24.684-0700	INFO	Ramen hub operator pod "ramen-hub-operator-78f5b7c4c4-l4zb6" is running in cluster "hub"
2025-07-11T05:53:24.685-0700	INFO	Ramen dr cluster operator pod "ramen-dr-cluster-operator-668789d94-vfcmd" is running in cluster "dr2"
2025-07-11T05:53:24.698-0700	INFO	Ramen dr cluster operator pod "ramen-dr-cluster-operator-668789d94-5tvng" is running in cluster "dr1"
--- PASS: TestValidation (0.16s)
    --- PASS: TestValidation/hub (0.02s)
    --- PASS: TestValidation/c2 (0.02s)
    --- PASS: TestValidation/c1 (0.04s)
PASS
[root@c57199v1 e2e]# ./run.sh -test.run TestDR
2025-07-11T05:53:34.583-0700	INFO	Using config file "config.yaml"
2025-07-11T05:53:34.584-0700	INFO	Using log file "ramen-e2e.log"
2025-07-11T05:53:34.592-0700	INFO	Using "hub" cluster name: "hub"
2025-07-11T05:53:34.633-0700	INFO	Detected "c1" managed cluster name: "dr1"
2025-07-11T05:53:34.667-0700	INFO	Detected "c2" managed cluster name: "dr2"
2025-07-11T05:53:34.668-0700	INFO	Using DeployTimeout: 5m0s
2025-07-11T05:53:34.668-0700	INFO	Using UneployTimeout: 5m0s
2025-07-11T05:53:34.668-0700	INFO	Using EnableTimeout: 5m0s
2025-07-11T05:53:34.668-0700	INFO	Using DisableTimeout: 5m0s
2025-07-11T05:53:34.668-0700	INFO	Using FailoverTimeout: 15m0s
2025-07-11T05:53:34.668-0700	INFO	Using RelocateTimeout: 15m0s
2025-07-11T05:53:34.668-0700	INFO	Using RetryInterval: 5s
=== RUN   TestDR
=== PAUSE TestDR
=== CONT  TestDR
2025-07-11T05:53:34.710-0700	INFO	Detected kubernetes distribution: "k8s"
2025-07-11T05:53:34.711-0700	INFO	Using namespaces: {RamenHubNamespace:ramen-system RamenDRClusterNamespace:ramen-system RamenOpsNamespace:ramen-ops ArgocdNamespace:argocd}
2025-07-11T05:53:34.812-0700	INFO	Validated clusters ["dr1", "dr2"] in DRPolicy "dr-policy"
2025-07-11T05:53:34.860-0700	INFO	Validated clusters ["dr1", "dr2"] in ClusterSet "default"
2025-07-11T05:53:34.920-0700	INFO	Created channel "e2e-gitops/https-github-com-ramendr-ocm-ramen-samples-git" in cluster "hub"
=== RUN   TestDR/appset-deploy-rbd-busybox
=== PAUSE TestDR/appset-deploy-rbd-busybox
=== RUN   TestDR/appset-deploy-cephfs-busybox
=== PAUSE TestDR/appset-deploy-cephfs-busybox
=== RUN   TestDR/subscr-deploy-rbd-busybox
=== PAUSE TestDR/subscr-deploy-rbd-busybox
=== RUN   TestDR/subscr-deploy-cephfs-busybox
=== PAUSE TestDR/subscr-deploy-cephfs-busybox
=== RUN   TestDR/disapp-deploy-rbd-busybox
=== PAUSE TestDR/disapp-deploy-rbd-busybox
=== RUN   TestDR/disapp-deploy-cephfs-busybox
=== PAUSE TestDR/disapp-deploy-cephfs-busybox
=== CONT  TestDR/appset-deploy-rbd-busybox
=== RUN   TestDR/appset-deploy-rbd-busybox/Deploy
2025-07-11T05:53:34.921-0700	INFO	appset-deploy-rbd-busybox	Deploying applicationset app "e2e-appset-deploy-rbd-busybox/busybox" in cluster "dr1"
=== CONT  TestDR/disapp-deploy-cephfs-busybox
=== RUN   TestDR/disapp-deploy-cephfs-busybox/Deploy
=== CONT  TestDR/disapp-deploy-rbd-busybox
=== RUN   TestDR/disapp-deploy-rbd-busybox/Deploy
=== CONT  TestDR/subscr-deploy-cephfs-busybox
=== RUN   TestDR/subscr-deploy-cephfs-busybox/Deploy
2025-07-11T05:53:34.922-0700	INFO	subscr-deploy-cephfs-busybox	Deploying subscription app "e2e-subscr-deploy-cephfs-busybox/busybox" in cluster "dr1"
=== CONT  TestDR/subscr-deploy-rbd-busybox
=== RUN   TestDR/subscr-deploy-rbd-busybox/Deploy
=== CONT  TestDR/appset-deploy-cephfs-busybox
2025-07-11T05:53:34.922-0700	INFO	subscr-deploy-rbd-busybox	Deploying subscription app "e2e-subscr-deploy-rbd-busybox/busybox" in cluster "dr1"
=== RUN   TestDR/appset-deploy-cephfs-busybox/Deploy
2025-07-11T05:53:34.922-0700	INFO	appset-deploy-cephfs-busybox	Deploying applicationset app "e2e-appset-deploy-cephfs-busybox/busybox" in cluster "dr1"
2025-07-11T05:53:35.038-0700	INFO	disapp-deploy-cephfs-busybox	Deploying discovered app "e2e-disapp-deploy-cephfs-busybox/busybox" in cluster "dr1"
2025-07-11T05:53:35.050-0700	INFO	disapp-deploy-rbd-busybox	Deploying discovered app "e2e-disapp-deploy-rbd-busybox/busybox" in cluster "dr1"
2025-07-11T05:53:53.656-0700	INFO	disapp-deploy-cephfs-busybox	Workload deployed
=== RUN   TestDR/disapp-deploy-cephfs-busybox/Enable
2025-07-11T05:53:53.688-0700	INFO	disapp-deploy-cephfs-busybox	Protecting workload "e2e-disapp-deploy-cephfs-busybox/busybox" in cluster "dr1"
2025-07-11T05:53:55.349-0700	INFO	appset-deploy-cephfs-busybox	Workload deployed
=== RUN   TestDR/appset-deploy-cephfs-busybox/Enable
2025-07-11T05:53:55.377-0700	INFO	appset-deploy-cephfs-busybox	Protecting workload "e2e-appset-deploy-cephfs-busybox/busybox" in cluster "dr1"
2025-07-11T05:53:55.728-0700	INFO	subscr-deploy-rbd-busybox	Workload deployed
=== RUN   TestDR/subscr-deploy-rbd-busybox/Enable
2025-07-11T05:53:55.729-0700	INFO	subscr-deploy-cephfs-busybox	Workload deployed
=== RUN   TestDR/subscr-deploy-cephfs-busybox/Enable
2025-07-11T05:53:55.760-0700	INFO	subscr-deploy-rbd-busybox	Protecting workload "e2e-subscr-deploy-rbd-busybox/busybox" in cluster "dr1"
2025-07-11T05:53:55.760-0700	INFO	subscr-deploy-cephfs-busybox	Protecting workload "e2e-subscr-deploy-cephfs-busybox/busybox" in cluster "dr1"
2025-07-11T05:53:58.753-0700	INFO	disapp-deploy-rbd-busybox	Workload deployed
=== RUN   TestDR/disapp-deploy-rbd-busybox/Enable
2025-07-11T05:53:58.815-0700	INFO	disapp-deploy-rbd-busybox	Protecting workload "e2e-disapp-deploy-rbd-busybox/busybox" in cluster "dr1"
2025-07-11T05:54:00.357-0700	INFO	appset-deploy-rbd-busybox	Workload deployed
=== RUN   TestDR/appset-deploy-rbd-busybox/Enable
2025-07-11T05:54:00.424-0700	INFO	appset-deploy-rbd-busybox	Protecting workload "e2e-appset-deploy-rbd-busybox/busybox" in cluster "dr1"
2025-07-11T05:56:02.281-0700	INFO	subscr-deploy-rbd-busybox	Workload protected
=== RUN   TestDR/subscr-deploy-rbd-busybox/Failover
2025-07-11T05:56:02.386-0700	INFO	subscr-deploy-rbd-busybox	Failing over workload "e2e-subscr-deploy-rbd-busybox/busybox" from cluster "dr1" to cluster "dr2"
2025-07-11T05:56:05.754-0700	INFO	disapp-deploy-rbd-busybox	Workload protected
=== RUN   TestDR/disapp-deploy-rbd-busybox/Failover
2025-07-11T05:56:06.065-0700	INFO	disapp-deploy-rbd-busybox	Failing over workload "e2e-disapp-deploy-rbd-busybox/busybox" from cluster "dr1" to cluster "dr2"
2025-07-11T05:56:07.424-0700	INFO	appset-deploy-rbd-busybox	Workload protected
=== RUN   TestDR/appset-deploy-rbd-busybox/Failover
2025-07-11T05:56:08.476-0700	INFO	appset-deploy-rbd-busybox	Failing over workload "e2e-appset-deploy-rbd-busybox/busybox" from cluster "dr1" to cluster "dr2"
2025-07-11T05:56:51.834-0700	INFO	appset-deploy-cephfs-busybox	Workload protected
=== RUN   TestDR/appset-deploy-cephfs-busybox/Failover
2025-07-11T05:56:52.286-0700	INFO	subscr-deploy-cephfs-busybox	Workload protected
=== RUN   TestDR/subscr-deploy-cephfs-busybox/Failover
2025-07-11T05:56:52.963-0700	INFO	appset-deploy-cephfs-busybox	Failing over workload "e2e-appset-deploy-cephfs-busybox/busybox" from cluster "dr1" to cluster "dr2"
2025-07-11T05:56:53.722-0700	INFO	subscr-deploy-cephfs-busybox	Failing over workload "e2e-subscr-deploy-cephfs-busybox/busybox" from cluster "dr1" to cluster "dr2"
2025-07-11T05:57:02.027-0700	INFO	disapp-deploy-cephfs-busybox	Workload protected
=== RUN   TestDR/disapp-deploy-cephfs-busybox/Failover
2025-07-11T05:57:02.144-0700	INFO	disapp-deploy-cephfs-busybox	Failing over workload "e2e-disapp-deploy-cephfs-busybox/busybox" from cluster "dr1" to cluster "dr2"
2025-07-11T06:01:53.566-0700	INFO	subscr-deploy-rbd-busybox	Workload failed over
=== RUN   TestDR/subscr-deploy-rbd-busybox/Relocate
2025-07-11T06:01:54.034-0700	INFO	subscr-deploy-rbd-busybox	Relocating workload "e2e-subscr-deploy-rbd-busybox/busybox" from cluster "dr2" to cluster "dr1"
2025-07-11T06:01:58.396-0700	INFO	disapp-deploy-rbd-busybox	Workload failed over
=== RUN   TestDR/disapp-deploy-rbd-busybox/Relocate
2025-07-11T06:01:58.516-0700	INFO	appset-deploy-rbd-busybox	Workload failed over
=== RUN   TestDR/appset-deploy-rbd-busybox/Relocate
2025-07-11T06:01:58.526-0700	INFO	disapp-deploy-rbd-busybox	Relocating workload "e2e-disapp-deploy-rbd-busybox/busybox" from cluster "dr2" to cluster "dr1"
2025-07-11T06:01:58.610-0700	INFO	appset-deploy-rbd-busybox	Relocating workload "e2e-appset-deploy-rbd-busybox/busybox" from cluster "dr2" to cluster "dr1"
2025-07-11T06:02:22.924-0700	INFO	subscr-deploy-cephfs-busybox	Workload failed over
=== RUN   TestDR/subscr-deploy-cephfs-busybox/Relocate
2025-07-11T06:02:22.974-0700	INFO	subscr-deploy-cephfs-busybox	Relocating workload "e2e-subscr-deploy-cephfs-busybox/busybox" from cluster "dr2" to cluster "dr1"
2025-07-11T06:03:51.096-0700	INFO	disapp-deploy-cephfs-busybox	Workload failed over
=== RUN   TestDR/disapp-deploy-cephfs-busybox/Relocate
2025-07-11T06:03:51.376-0700	INFO	disapp-deploy-cephfs-busybox	Relocating workload "e2e-disapp-deploy-cephfs-busybox/busybox" from cluster "dr2" to cluster "dr1"
2025-07-11T06:06:09.017-0700	INFO	appset-deploy-cephfs-busybox	Workload failed over
=== RUN   TestDR/appset-deploy-cephfs-busybox/Relocate
2025-07-11T06:06:09.061-0700	INFO	appset-deploy-cephfs-busybox	Relocating workload "e2e-appset-deploy-cephfs-busybox/busybox" from cluster "dr2" to cluster "dr1"
2025-07-11T06:06:13.315-0700	INFO	subscr-deploy-rbd-busybox	Workload relocated
=== RUN   TestDR/subscr-deploy-rbd-busybox/Disable
2025-07-11T06:06:13.346-0700	INFO	subscr-deploy-rbd-busybox	Unprotecting workload "e2e-subscr-deploy-rbd-busybox/busybox" in cluster "dr1"
2025-07-11T06:06:20.421-0700	INFO	disapp-deploy-rbd-busybox	Workload relocated
=== RUN   TestDR/disapp-deploy-rbd-busybox/Disable
2025-07-11T06:06:20.449-0700	INFO	disapp-deploy-rbd-busybox	Unprotecting workload "e2e-disapp-deploy-rbd-busybox/busybox" in cluster "dr1"
2025-07-11T06:06:30.730-0700	INFO	subscr-deploy-rbd-busybox	Workload unprotected
=== RUN   TestDR/subscr-deploy-rbd-busybox/Undeploy
2025-07-11T06:06:30.803-0700	INFO	subscr-deploy-rbd-busybox	Undeploying subscription app "e2e-subscr-deploy-rbd-busybox/busybox" in cluster "dr1"
2025-07-11T06:06:42.686-0700	INFO	subscr-deploy-rbd-busybox	Workload undeployed
2025-07-11T06:06:44.005-0700	INFO	appset-deploy-rbd-busybox	Workload relocated
=== RUN   TestDR/appset-deploy-rbd-busybox/Disable
2025-07-11T06:06:44.078-0700	INFO	appset-deploy-rbd-busybox	Unprotecting workload "e2e-appset-deploy-rbd-busybox/busybox" in cluster "dr1"
2025-07-11T06:07:18.198-0700	INFO	disapp-deploy-rbd-busybox	Workload unprotected
=== RUN   TestDR/disapp-deploy-rbd-busybox/Undeploy
2025-07-11T06:07:18.198-0700	INFO	disapp-deploy-rbd-busybox	Undeploying discovered app "e2e-disapp-deploy-rbd-busybox/busybox" in clusters "dr1" and "dr2"
2025-07-11T06:07:35.946-0700	INFO	disapp-deploy-rbd-busybox	Workload undeployed
2025-07-11T06:10:44.565-0700	INFO	subscr-deploy-cephfs-busybox	Workload relocated
=== RUN   TestDR/subscr-deploy-cephfs-busybox/Disable
2025-07-11T06:10:44.590-0700	INFO	subscr-deploy-cephfs-busybox	Unprotecting workload "e2e-subscr-deploy-cephfs-busybox/busybox" in cluster "dr1"
2025-07-11T06:10:50.043-0700	INFO	appset-deploy-rbd-busybox	Workload unprotected
=== RUN   TestDR/appset-deploy-rbd-busybox/Undeploy
2025-07-11T06:10:50.077-0700	INFO	appset-deploy-rbd-busybox	Undeploying applicationset app "e2e-appset-deploy-rbd-busybox/busybox" in cluster "dr1"
2025-07-11T06:11:08.841-0700	INFO	appset-deploy-rbd-busybox	Workload undeployed
2025-07-11T06:11:43.611-0700	INFO	subscr-deploy-cephfs-busybox	Workload unprotected
=== RUN   TestDR/subscr-deploy-cephfs-busybox/Undeploy
2025-07-11T06:11:43.645-0700	INFO	subscr-deploy-cephfs-busybox	Undeploying subscription app "e2e-subscr-deploy-cephfs-busybox/busybox" in cluster "dr1"
2025-07-11T06:12:01.438-0700    INFO    subscr-deploy-cephfs-busybox	Workload undeployed
2025-07-11T06:12:47.094-0700	INFO	appset-deploy-cephfs-busybox	Workload relocated
=== RUN   TestDR/appset-deploy-cephfs-busybox/Disable
2025-07-11T06:12:47.123-0700	INFO	appset-deploy-cephfs-busybox	Unprotecting workload "e2e-appset-deploy-cephfs-busybox/busybox" in cluster "dr1"
2025-07-11T06:17:47.095-0700	ERROR	appset-deploy-cephfs-busybox	Failed to disable protection for workload: DRPlacementControl "argocd/appset-deploy-cephfs-busybox" not deleted in cluster "hub": context deadline exceeded
2025-07-11T06:17:56.953-0700	INFO	disapp-deploy-cephfs-busybox	Workload relocated
=== RUN   TestDR/disapp-deploy-cephfs-busybox/Disable
2025-07-11T06:17:57.008-0700	INFO	disapp-deploy-cephfs-busybox	Unprotecting workload "e2e-disapp-deploy-cephfs-busybox/busybox" in cluster "dr1"
2025-07-11T06:18:27.178-0700	INFO	disapp-deploy-cephfs-busybox	Workload unprotected
=== RUN   TestDR/disapp-deploy-cephfs-busybox/Undeploy
2025-07-11T06:18:27.179-0700	INFO	disapp-deploy-cephfs-busybox	Undeploying discovered app "e2e-disapp-deploy-cephfs-busybox/busybox" in clusters "dr1" and "dr2"
2025-07-11T06:18:43.216-0700	INFO	disapp-deploy-cephfs-busybox	Workload undeployed
2025-07-11T06:18:43.264-0700	INFO	Deleted channel "e2e-gitops/https-github-com-ramendr-ocm-ramen-samples-git" in cluster "hub"
--- FAIL: TestDR (8.58s)
    --- PASS: TestDR/subscr-deploy-rbd-busybox (787.76s)
        --- PASS: TestDR/subscr-deploy-rbd-busybox/Deploy (20.81s)
        --- PASS: TestDR/subscr-deploy-rbd-busybox/Enable (126.55s)
        --- PASS: TestDR/subscr-deploy-rbd-busybox/Failover (351.29s)
        --- PASS: TestDR/subscr-deploy-rbd-busybox/Relocate (259.75s)
        --- PASS: TestDR/subscr-deploy-rbd-busybox/Disable (17.42s)
        --- PASS: TestDR/subscr-deploy-rbd-busybox/Undeploy (11.96s)
    --- PASS: TestDR/disapp-deploy-rbd-busybox (841.02s)
        --- PASS: TestDR/disapp-deploy-rbd-busybox/Deploy (23.83s)
        --- PASS: TestDR/disapp-deploy-rbd-busybox/Enable (127.00s)
        --- PASS: TestDR/disapp-deploy-rbd-busybox/Failover (352.64s)
        --- PASS: TestDR/disapp-deploy-rbd-busybox/Relocate (262.02s)
        --- PASS: TestDR/disapp-deploy-rbd-busybox/Disable (57.78s)
        --- PASS: TestDR/disapp-deploy-rbd-busybox/Undeploy (17.75s)
    --- PASS: TestDR/appset-deploy-rbd-busybox (1053.92s)
        --- PASS: TestDR/appset-deploy-rbd-busybox/Deploy (25.44s)
        --- PASS: TestDR/appset-deploy-rbd-busybox/Enable (127.07s)
        --- PASS: TestDR/appset-deploy-rbd-busybox/Failover (351.09s)
        --- PASS: TestDR/appset-deploy-rbd-busybox/Relocate (285.49s)
        --- PASS: TestDR/appset-deploy-rbd-busybox/Disable (246.04s)
        --- PASS: TestDR/appset-deploy-rbd-busybox/Undeploy (18.80s)
    --- PASS: TestDR/subscr-deploy-cephfs-busybox (1106.52s)
        --- PASS: TestDR/subscr-deploy-cephfs-busybox/Deploy (20.81s)
        --- PASS: TestDR/subscr-deploy-cephfs-busybox/Enable (176.56s)
        --- PASS: TestDR/subscr-deploy-cephfs-busybox/Failover (330.64s)
        --- PASS: TestDR/subscr-deploy-cephfs-busybox/Relocate (501.64s)
        --- PASS: TestDR/subscr-deploy-cephfs-busybox/Disable (59.05s)
        --- PASS: TestDR/subscr-deploy-cephfs-busybox/Undeploy (17.83s)
    --- FAIL: TestDR/appset-deploy-cephfs-busybox (1452.17s)
        --- PASS: TestDR/appset-deploy-cephfs-busybox/Deploy (20.43s)
        --- PASS: TestDR/appset-deploy-cephfs-busybox/Enable (176.49s)
        --- PASS: TestDR/appset-deploy-cephfs-busybox/Failover (557.18s)
        --- PASS: TestDR/appset-deploy-cephfs-busybox/Relocate (398.08s)
        --- FAIL: TestDR/appset-deploy-cephfs-busybox/Disable (300.00s)
    --- PASS: TestDR/disapp-deploy-cephfs-busybox (1508.29s)
        --- PASS: TestDR/disapp-deploy-cephfs-busybox/Deploy (18.74s)
        --- PASS: TestDR/disapp-deploy-cephfs-busybox/Enable (188.37s)
        --- PASS: TestDR/disapp-deploy-cephfs-busybox/Failover (409.07s)
        --- PASS: TestDR/disapp-deploy-cephfs-busybox/Relocate (845.86s)
        --- PASS: TestDR/disapp-deploy-cephfs-busybox/Disable (30.23s)
        --- PASS: TestDR/disapp-deploy-cephfs-busybox/Undeploy (16.04s)
FAIL
```

Fixes: #2180

nirs reviewed (commented):
Please remove root and duplicate signed offs.

```
Signed-off-by: pruthvitd <prd@redhat.com>
Signed-off-by: root <root@c57199v1.fyre.ibm.com>
Signed-off-by: pruthvitd <prd@redhat.com>
Signed-off-by: root <root@c57199v1.fyre.ibm.com>
````

nirs commented on go.mod:
```diff
@@ -1,10 +1,10 @@
-// Recommended version: latest go 1.23 release.
-toolchain go1.23.7
+// Recommended version: latest go 1.24 release.
+toolchain go1.24.1
```
1.24.4 is the latest version.

pruthvitd replied:
Updated to the latest go1.24.5 version

nirs commented on go.mod:
```diff
@@ -1,10 +1,10 @@
 
 // Required minimum version, must be available in downstream builders.
-go 1.23.5
+go 1.24
```
This must be x.y.z format, so 1.24.0

pruthvitd replied:
Updated as per the review comment. Thanks

nirs commented on hack/install-golangci-lint.sh:
```diff
@@ -4,7 +4,7 @@ set -e
 
 source_url="https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh"
-required_version="1.62.0"
+required_version="1.64.8"
```
If this works we are fine, but we really need to update to 2.x.

pruthvitd replied:
> If this works we are fine, but we really need to update to 2.x.

'golangci-lint' is updated using this command:

`
which golangci-lint
rm -f $(which golangci-lint)
go install github.com/golangci/golangci-lint/cmd/golangci-lint@latest
`

Updating golangci-lint involves changes related to linter config, will commit separately for linter upgrade.

nirs reviewed (commented):
We need to update also e2e api modules.

nirs commented on hack/install-golangci-lint.sh:
```diff
@@ -4,7 +4,7 @@ set -e
 
 source_url="https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh"
-required_version="1.62.0"
+required_version="1.64.8"
```
This does not work, check the CI:

```
Running [/home/runner/golangci-lint-1.62.0-linux-amd64/golangci-lint config path] in [/home/runner/work/ramen/ramen] ...
  Running [/home/runner/golangci-lint-1.62.0-linux-amd64/golangci-lint run --path-prefix=.] in [/home/runner/work/ramen/ramen] ...
  Error: can't load config: the Go language version (go1.23) used to build golangci-lint is lower than the targeted Go version (1.24.4)
  Failed executing command with error: can't load config: the Go language version (go1.23) used to build golangci-lint is lower than the targeted Go version (1.24.4)
```

I think we must start by upgrading to golangci-lint 2.x - before updating to go 1.24. I'm running this locally for a while with latest go 1.24 and it works fine.

We need to migrate the golangci.yaml configuration to version 2, see the golangci-lint migrate command.

This is the local configuration I'm using:

```yaml
% cat .golangci-v2.yaml
version: "2"
linters:
  default: none
  enable:
    - asasalint
    - asciicheck
    - bidichk
    - bodyclose
    - cyclop
    - decorder
    - dogsled
    - dupl
    - durationcheck
    - errcheck
    - errchkjson
    - exhaustive
    - forbidigo
    - funlen
    - gocognit
    - gocritic
    - gocyclo
    - gomodguard
    - goprintffuncname
    - gosec
    - govet
    - grouper
    - importas
    - ineffassign
    - interfacebloat
    - lll
    - loggercheck
    - maintidx
    - makezero
    - misspell
    - mnd
    - nestif
    - nilerr
    - nlreturn
    - noctx
    - nosprintfhostport
    - prealloc
    - predeclared
    - promlinter
    - reassign
    - revive
    - rowserrcheck
    - sqlclosecheck
    - staticcheck
    - testpackage
    - thelper
    - tparallel
    - unconvert
    - unparam
    - unused
    - usestdlibvars
    - wastedassign
    - whitespace
    - wsl
  settings:
    errcheck:
      check-type-assertions: true
      check-blank: true
    gocognit:
      min-complexity: 15
    gocyclo:
      min-complexity: 15
    goheader:
      template: |
        SPDX-FileCopyrightText: The RamenDR authors
        SPDX-License-Identifier: Apache-2.0
    misspell:
      locale: US
    promlinter:
      strict: true
    revive:
      severity: error
      rules:
        - name: blank-imports
        - name: context-as-argument
        - name: context-keys-type
        - name: dot-imports
        - name: error-return
        - name: error-strings
        - name: error-naming
        - name: exported
        - name: increment-decrement
        - name: var-naming
        - name: var-declaration
        - name: package-comments
        - name: range
        - name: receiver-naming
        - name: time-naming
        - name: unexported-return
        - name: indent-error-flow
        - name: errorf
        - name: empty-block
        - name: superfluous-else
        - name: unreachable-code
        - name: redefines-builtin-id
        - name: atomic
        - name: constant-logical-expr
        - name: unnecessary-stmt
        - name: get-return
        - name: modifies-parameter
        - name: modifies-value-receiver
        - name: range-val-in-closure
        - name: waitgroup-by-value
        - name: call-to-gc
        - name: duplicated-imports
        - name: unhandled-error
    wsl:
      allow-trailing-comment: true
      force-err-cuddling: true
  exclusions:
    generated: lax
    presets:
      - comments
      - common-false-positives
      - legacy
      - std-error-handling
    rules:
      - linters:
          - revive
        text: should not use dot imports
        source: ginkgo|gomega
      - linters:
          - funlen
        source: ^func Test
      - linters:
          - lll
        source: ^//
    paths:
      - third_party$
      - builtin$
      - examples$
formatters:
  enable:
    - gofmt
    - gofumpt
  exclusions:
    generated: lax
    paths:
      - third_party$
      - builtin$
      - examples$
```

The automatic migration drops all comments (and commented out linters). We probably want to keep them in official configuration.

pruthvitd replied:
After migrating golangci-lint config and resolving errors:
```
# make lint
testbin/golangci-lint config verify --config=./.golangci.yaml
cd e2e && ../testbin/golangci-lint run ./... --config=../.golangci.yaml
WARN [runner/nolint_filter] Found unknown linters in //nolint directives: golint, stylecheck 
0 issues.
cd api && ../testbin/golangci-lint run ./... --config=../.golangci.yaml
0 issues.
testbin/golangci-lint run ./... --config=./.golangci.yaml
0 issues.
hack/pre-commit.sh
/tmp/tool-errors-9rAVSD
=====  mdl =====


=====  shellcheck  =====


=====  yamllint  =====

```

pruthvitd replied:
wsl is deprecated hence updated to wsl_v5 and commented/disabled static-check and var-naming linters.

pruthvitd commented:
> Please remove root and duplicate signed offs.
> 
> ```
> Signed-off-by: pruthvitd <prd@redhat.com>
> Signed-off-by: root <root@c57199v1.fyre.ibm.com>
> Signed-off-by: pruthvitd <prd@redhat.com>
> Signed-off-by: root <root@c57199v1.fyre.ibm.com>
> ```



> We need to update also e2e api modules.

Done

nirs commented:
> > @pruthvitd We need to upgrade to golangci-lint 2.x before we can upgrade go version.
> 
> I see more than 50 linter errors with golangci-lint v2.2.2

Can you post a PR upgrading golangci-lint to latest version?

We can start by disabling linter settings which have changes since v1. We already have some linters which were commented out.

We can also split the work easily.

pruthvitd commented:
> We need to update also e2e api modules.

Done

pruthvitd commented:
> > > @pruthvitd We need to upgrade to golangci-lint 2.x before we can upgrade go version.
> > 
> > 
> > I see more than 50 linter errors with golangci-lint v2.2.2
> 
> Can you post a PR upgrading golangci-lint to latest version?
> 
> We can start by disabling linter settings which have changes since v1. We already have some linters which were commented out.
> 
> We can also split the work easily.

All the linter errors are addressed and updated yamllint, ShellCheck and mdl versions.

pruthvitd commented:
> > > This is too much. We are not saving time by doing everything in one commit or one pr. This is a good way to wasted time and delay the changes. We want to make progress in small steps - easy to review, easy to get right, easy to test.
> > > Please separate the golangci-lint upgrade and lint fixes.
> > > I'm not sure about the other changes, not sure how Kubernetes 0.30.0 is related.
> > > Update mdl, shellcheck and yamllint should not be related, but if we need to update them it should be in a separate commit or pr.
> > 
> > 
> > @nirs This was the reason I wanted the linter upgrade to be different PR, since you mentioned about upgrading linter these changes got introduced.
> 
> When we upgrade the linter, we many need to fix the code to match the linter, or we need to disable new features in the linter to match the code. It seems that you did the first successfuly, so can have a pr updating golangci-lint with the code changes to make it pass. The next pr can update go version, which seems to be block by the lint issues.

PR for linter upgrade: https://github.com/RamenDR/ramen/pull/2174
